### PR TITLE
[WIP] feat: inbuilt dialer including call history

### PR DESCRIPTION
**Pull request by Bnyro**
TODO:
- [X] a button to mute the mic and one to turn the normal speakers on for audio output
- [X] Dual sim support
- [X] a notification to get back to the current call
- [x] In call dialpad
- [ ] Proximity Sensor support
- [ ] testing !!!

NOTES:
- the minSdk bump is necessary because lower Android versions don't allow to use the InCallService

### CODE REVIEW

**Comment by M00NJ** [comment]
Some things I noticed while doing some preliminary testing:

1. Selecting a contact from the contacts tab and clicking on dial doesn't work (nothing happens)
2. When the phone is not picked up and the call goes to the mailbox, it automatically switches to the AOSP phone app an the call is continued there (really strange)
3. There's no way to delete calls (swipe to dismiss and a 'clear all' button would make sence here)

Some other things we should considers:
- Implementing search in the phone tab as well 
  (ATM with the dial button not working there's no way to select a contact that you haven't called recently, so you'll have to know the number and dial it.)
- Use a list instead of cards
  (I mentioned this already for the contacts tab, but it applies even more so here. If we're listing items we should use a list 😅. So many cards on screen make the UI look cramped IMO and waste vertical space, where there might fit another contact on screen.)

**Comment by Gamer750** [comment]
Is the dialer included in the next release

File: app/src/main/java/com/bnyro/contacts/services/CallService.kt [hunk 1]
```diff
@@ -50,32 +61,93 @@ class CallService : InCallService() {
...
+
+    private fun getContactPhotoThumbnail(uri: String): Bitmap? {
+        val contentResolver = this.contentResolver
+        val photoThumbnailUri = Uri.parse(uri)
+        val assetFileDescriptor =
+            contentResolver.openAssetFileDescriptor(photoThumbnailUri, "r") ?: return null
+        val fileDescriptor = assetFileDescriptor.fileDescriptor
+        val bitmap = BitmapFactory.decodeFileDescriptor(fileDescriptor)
+        assetFileDescriptor.close()
+        return bitmap
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/contacts/services/CallService.kt:122]: We probably already do almost the same in the `DeviceContactsRepository`, so we should move this tothe `ContactHelper` object or so.

File: app/src/main/java/com/bnyro/contacts/services/CallService.kt [hunk 2]
```diff
@@ -50,32 +61,93 @@ class CallService : InCallService() {
...
+                ?: return null to null
+        var contactName: String? = null
+        var contactThumbnail: String? = null
+
+        query.use { cursor ->
+            if (cursor.moveToFirst()) {
+                val nameIndex = cursor.getColumnIndex(PhoneLookup.DISPLAY_NAME)
+                val thumbnailIndex = cursor.getColumnIndex(PhoneLookup.PHOTO_THUMBNAIL_URI)
+                contactName = cursor.getString(nameIndex)
+                contactThumbnail = cursor.getString(thumbnailIndex)
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/contacts/services/CallService.kt:107]:
```suggestion
                val nameIndex = cursor.stringValue(PhoneLookup.DISPLAY_NAME)
                val thumbnailIndex = cursor.stringValue(PhoneLookup.PHOTO_THUMBNAIL_URI)
```

File: app/src/main/java/com/bnyro/contacts/services/CallService.kt [hunk 3]
```diff
@@ -50,32 +61,93 @@ class CallService : InCallService() {
...
+            cr.query(
+                uri,
+                arrayOf(PhoneLookup.DISPLAY_NAME, PhoneLookup.PHOTO_THUMBNAIL_URI),
+                null,
+                null,
+                null
+            )
+                ?: return null to null
+        var contactName: String? = null
+        var contactThumbnail: String? = null
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/contacts/services/CallService.kt:100]: We actually don't need these variables - they can be inlined to line 106/107 instead

File: app/src/main/java/com/bnyro/contacts/services/CallService.kt [hunk 4]
```diff
@@ -50,32 +61,93 @@ class CallService : InCallService() {
...
+    private fun buildNotification(
+        call: Call,
+        callerNumber: String,
+        callerName: String?,
+        callersPhoto: Bitmap?
+    ): Notification {
         val collapsedView = RemoteViews(packageName, R.layout.call_notification).apply {
-            setTextViewText(R.id.notification_caller_name, "TODO: Get Contact Name")
-            setTextViewText(R.id.notification_call_status, CallManager.callerDisplayNumber)
+            setTextViewText(R.id.notification_caller_name, callerName ?: "Unknown Number")
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/contacts/services/CallService.kt:139]:
```suggestion
            setTextViewText(R.id.notification_caller_name, callerName ?: getString(R.string.unknown_number))
```

**Comment by SuhasDissa** [comment]
I think it will be easier if we could merge this PR now and work on the other features after merging.
Integrating contacts with the dialer will probably require some refactoring, which could cause merge conflicts.

Apart from that, I believe that the dialer is stable enough to mark as an experimental opt-in feature.

**Comment by Bnyro** [comment]
I agree on that, let's merge :+1:

Before creating a new release we can just make a preference to opt-in for using it.

**Pull request merged**